File: web/src/lib/revolutionRange.ts
export const REV_MIN = 0.5
export const REV_MAX = 4
export const REV_STEP = 0.5

/**
 * A full-range selection (REV_MIN..REV_MAX) encodes to an empty list, matching
 * the existing "empty AllowedRevolutions = no restriction" backend semantics.
 *
 * Assumes min <= max and both are already snapped to the 0.5 grid within
 * [REV_MIN, REV_MAX] — callers (the range-slider components) guarantee this
 * before invoking it.
 */
export function encodeRevolutionRange(min: number, max: number): number[] {
  if (min <= REV_MIN && max >= REV_MAX) return []
  const steps = Math.round((max - min) / REV_STEP)
  const result: number[] = []
  for (let i = 0; i <= steps; i++) {
    result.push(Math.round((min + i * REV_STEP) * 2) / 2)
  }
  return result
}

export function decodeRevolutionRange(list: number[]): { min: number; max: number } {
  if (list.length === 0) return { min: REV_MIN, max: REV_MAX }
  return { min: Math.min(...list), max: Math.max(...list) }
}

File: mobile/lib/core/models/revolution_range.dart
const double kRevolutionRangeMin = 0.5;
const double kRevolutionRangeMax = 4.0;
const double kRevolutionRangeStep = 0.5;

/// A full-range selection (kRevolutionRangeMin..kRevolutionRangeMax) encodes
/// to an empty list, matching the existing "empty AllowedRevolutions = no
/// restriction" backend semantics.
List<double> encodeRevolutionRange(double min, double max) {
  if (min <= kRevolutionRangeMin && max >= kRevolutionRangeMax) return [];
  final steps = ((max - min) / kRevolutionRangeStep).round();
  return [
    for (var i = 0; i <= steps; i++)
      ((min + i * kRevolutionRangeStep) * 2).round() / 2,
  ];
}

({double min, double max}) decodeRevolutionRange(List<double> list) {
  if (list.isEmpty) {
    return (min: kRevolutionRangeMin, max: kRevolutionRangeMax);
  }
  var lo = list.first;
  var hi = list.first;
  for (final v in list) {
    if (v < lo) lo = v;
    if (v > hi) hi = v;
  }
  return (min: lo, max: hi);
}
